File: Data Structures & Algorithms/trapping-rain-water/submission-0.java
class Solution {
    public int trap(int[] height) {
        int n=height.length;
        int l=0,r=n-1;
        int lm=height[0],rm=height[n-1];
        int res=0;
        while(l<r){
            if(lm<rm){
                l++;
                lm=Math.max(lm,height[l]);
                res+=lm-height[l];
            }else{
                r--;
                rm=Math.max(rm,height[r]);
                res+=rm-height[r];
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        st=[]
        maxarea=0
        for i in range(len(heights)+1):
            curr = 0 if i == len(heights) else heights[i]
            while st and curr<heights[st[-1]]:
                height=heights[st.pop()]
                left= -1 if not st else st[-1]
                width=i-left-1
                area=height*width
                maxarea=max(area,maxarea)
            if i<len(heights):st.append(i)
        return maxarea

